zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestFederatedZkClient.java

```diff
@@ -0,0 +1,252 @@
+package org.apache.helix.zookeeper.impl.client;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.helix.msdcommon.datamodel.TrieRoutingData;
+import org.apache.helix.msdcommon.exception.InvalidRoutingDataException;
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+import org.apache.helix.zookeeper.impl.ZkTestBase;
+import org.apache.helix.zookeeper.zkclient.IZkStateListener;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.Op;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.ZooDefs;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class TestFederatedZkClient extends ZkTestBase {
```

Comment:
Great job on writing tests cases to cover all unsupported operations! This test looks great.

One more test case if I may suggest adding for FederatedZkClient is doing two writes to two different paths (path A and path B) where the two paths map to different ZK realms.

First, instrument the MetadataStoreRoutingData so that paths A and B have sharding keys that map to two different ZK realms. Then use the same FederatedZkClient to write to paths A and B.

Then create two single-realm ZkClients connected to the two realms respectively, and check that the data/ZNode you wrote to path A is in one ZK realm and not in the other, and vice versa.

Do you think that would be a good test case that tests FederatedZkClient's multi-realm functionality? :)

Reply:
This is correct. We would definitely need this test to very it is actual a multi realm client. I’ve considered it and it is in progress. Because in the previous version there is only one zk server running, I could not  run multi realms. I was planning to start a zk server just for this test but was waiting to see if we want to start one more zk in zk tests base. I will push a test. Thanks for the review and suggestion!

Reply:
@pkuwm 
Our ZkTestBase already starts multiple ZKs and you could reference them in _zkServerMap :)

You could look in our pom.xml file and search for "multiZk" keyword. I believe I have it set at 3. And you could use a protected variable `_numZk` to check how many ZKs have been started. That should make writing a test case for this much easier (start by doing a check - `if (_numZk > 1)` )

Reply:
@narendly I've noticed that multi-zk execution is included in pom. But actually I not clear why we are running two times of tests. And actually in Intellij the test runs the default-tests but not multi-zk (maybe I need to config my Intellij).
Anyway, I've added a test for multi realm.

Reply:
I don't think IntelliJ supports this - you could run it in command line with maven directly. That way, it will execute twice.

Reply:
`if (_numZk > 1)` to check multi-zk env in each multi-realm test doesn't look natural to me. Why can we not just keep one execution which just initially starts multi ZK? Does it hurt the default tests which don't need multi-zk env?

Reply:
We could make an exception to tests written for ZkClient APIs, but in general, we want to run all tests in two environments: single-realm and multi-realm. 

If we only keep one execution with multiple ZKs, then we lose test coverage for the single-realm case.